syntax = "proto2";
package geography.proto;

import "games/market/proto/goods.proto";
import "games/market/proto/market.proto";
import "games/industry/proto/industry.proto";
import "util/proto/object_id.proto";

// A unit area, able to hold one production chain.
message Field {
  // Progress of the production chain currently underway.
  optional industry.proto.Progress progress = 1;
  // Fixed equipment for this field.
  optional market.proto.Container fixed_capital = 2;
  // Terrain type.
  optional industry.proto.LandType land_type = 3;
  // Raw materials, for example ores.
  optional market.proto.Container resources = 4;
  // ID of the field's owner.
  optional uint64 owner_id = 5;
  // Identifier string for debug logging.
  optional string name = 6;
}

message ResourceLimits {
  // The maximum amount of each raw material in fields in this area.
  optional market.proto.Container maximum = 1;
  // Recovery speed while a production chain is in progress.
  optional market.proto.Container recovery = 2;
  // Same while there is no production.
  optional market.proto.Container fallow_recovery = 3;
}

// A collection of fields close enough together to be treated as a unit, for
// example, to have a single market.
message Area {
  repeated Field fields = 1;
  optional ResourceLimits limits = 2;
  repeated uint64 pop_ids = 3;
  optional market.proto.MarketProto market = 4;
  optional util.proto.ObjectId area_id = 5;
}

// A series of steps by which a Field can be converted from one LandType to
// another.
message Transition {
  optional industry.proto.LandType source = 1;
  optional industry.proto.LandType target = 2;

  // The fixed capital the target will end up with, for example houses.
  optional market.proto.Container final_fixed_capital = 3;

  // Useful byproducts, e.g. timber from converting a forest.
  optional market.proto.Container resources_harvested = 4;

  // The amount of fixed capital that requires one additional removal step.
  optional market.proto.Container step_fixed_capital = 5;

  // Resources required for one step.
  optional market.proto.Container step_input = 6;

  // Number of steps in the absence of fixed capital.
  optional int32 steps = 7;
}

// Types of connection.
enum ConnectionType {
  CT_OPEN = 0;
  CT_FOREST = 1;
  CT_MARSH = 2;
  CT_HILLS = 3;
  CT_MOUNTAIN = 4;
  CT_DESERT = 5;
  CT_RIVERINE = 6;
  CT_COASTAL = 7;
  CT_OPEN_SEA = 8;
}

// A connection between two areas.
message Connection {
  optional ConnectionType type = 1;
  optional uint64 id = 2;
  optional uint64 distance_u = 3;
  optional uint64 width_u = 4;
  optional int32 ruggedness = 5;
  optional int32 chokepoints = 6;

  optional util.proto.ObjectId a_area_id = 7;
  optional util.proto.ObjectId z_area_id = 8;
  // TODO: Infrastructure, vulnerability.
}

message Location {
  optional util.proto.ObjectId a_area_id = 1;
  optional util.proto.ObjectId z_area_id = 2;
  optional uint64 progress_u = 3;
  optional uint64 connection_id = 4;
}

